syntax = "proto3";

package masterworker;

service WorkerService {
  rpc map(Shard) returns (MapResult) {}
  rpc reduce(Region) returns (ReduceResult) {}
}

message ShardComponent {
  string file_path = 1;
  int32 start = 2;
  int32 size = 3;
}

message Shard {
  int32 id = 1;
  repeated ShardComponent components = 2;
}

message Region {
  int32 id = 1;
  repeated string file_paths = 2;
}

message MapResult {
  string worker_ip_addr_port = 1; // worker address as id
  repeated string file_path = 2;
}

message ReduceResult {
  string worker_ip_addr_port = 1; // worker address as id
  string file_path = 2;
}
